### PR TITLE
[RPC API Improvement 2/n] - Event query pagination and merge all getEvents* methods

### PR DESCRIPTION
**Pull request by patrickkuo**
# RPC API improvement part 2 : Event query pagination

## Changes:

* Merged all getEvents* methods and use EventQuery criteria object instead of separated methods.
```
pub enum EventQuery {
    /// Return all events.
    All,
    /// Return events emitted by the given transaction.
    Transaction(
        ///digest of the transaction, as base-64 encoded string
        TransactionDigest,
    ),
    /// Return events emitted in a specified Move module
    MoveModule {
        /// the Move package ID
        package: ObjectID,
        /// the module name
        module: String,
    },
    /// Return events with the given move event struct name
    MoveEvent(
        /// the event struct name type, e.g. `0x2::devnet_nft::MintNFTEvent` or `0x2::SUI::test_foo<address, vector<u8>>` with type params
        String,
    ),
    EventType(EventType),
    /// Query by sender address.
    Sender(SuiAddress),
    /// Query by recipient address.
    Recipient(Owner),
    /// Return events associated with the given object
    Object(ObjectID),
    /// Return events emitted in [start_time, end_time] interval
    TimeRange {
        /// left endpoint of time interval, inclusive
        start_time: u64,
        /// right endpoint of time interval, exclusive
        end_time: u64,
    },
}
```

* Added a unique ID to Event, the ID is generated automatically by the database during persistent, the ID is being used for pagination.
* Parse Move event fields.
* Capped RPC query page size limit to 1000.
* Added event query method to TS SDK.

## Examples
1, Querying all event
Request:
```
curl --location --request POST '127.0.0.1:9000' \
--header 'Content-Type: application/json' \
--data-raw '{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "sui_getEvents",
  "params": ["All", null, 1, "Descending"]
}'
```
Response:
```
{
    "jsonrpc": "2.0",
    "result": {
        "data": [
            {
                "id": 2,
                "timestamp": 1666012833101,
                "txDigest": "/fU42IZ6Bp0mRBwKfOO1iDxD9HtjBNuiQ+9TdNrEVWs=",
                "event": {
                    "moveEvent": {
                        "packageId": "0x0000000000000000000000000000000000000002",
                        "transactionModule": "devnet_nft",
                        "sender": "0xdac9483c5334c3189a740ad1f40fba71327aac5c",
                        "type": "0x2::devnet_nft::MintNFTEvent",
                        "fields": {
                            "creator": "0xdac9483c5334c3189a740ad1f40fba71327aac5c",
                            "name": "Example NFT",
                            "object_id": "0x18b41c66bfc56cc89833adb2fa11efa2f51e1bb7"
                        },
                        "bcs": "GLQcZr/FbMiYM62y+hHvovUeG7fayUg8UzTDGJp0CtH0D7pxMnqsXAtFeGFtcGxlIE5GVA=="
                    }
                }
            }
        ],
        "nextCursor": 1
    },
    "id": 1
}
```
2, Query by event type
request:
```
curl --location --request POST '127.0.0.1:9000' \
--header 'Content-Type: application/json' \
--data-raw '{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "sui_getEvents",
  "params": [{"EventType": "NewObject"}, null, null, "Descending"]
}'
```
response:
```
{
    "jsonrpc": "2.0",
    "result": {
        "data": [
            {
                "id": 1,
                "timestamp": 1666012833101,
                "txDigest": "/fU42IZ6Bp0mRBwKfOO1iDxD9HtjBNuiQ+9TdNrEVWs=",
                "event": {
                    "newObject": {
                        "packageId": "0x0000000000000000000000000000000000000002",
                        "transactionModule": "devnet_nft",
                        "sender": "0xdac9483c5334c3189a740ad1f40fba71327aac5c",
                        "recipient": {
                            "AddressOwner": "0xdac9483c5334c3189a740ad1f40fba71327aac5c"
                        },
                        "objectId": "0x18b41c66bfc56cc89833adb2fa11efa2f51e1bb7"
                    }
                }
            }
        ],
        "nextCursor": null
    },
    "id": 1
}
```

## TODOs
* https://github.com/MystenLabs/sui/issues/5501

### CODE REVIEW

File: sdk/typescript/src/providers/provider.ts
Comment (666lcz) [inline sdk/typescript/src/providers/provider.ts:200]: Since we cut the commit for DevNet 0.12.0, which does not include this commit, and we would like to do an SDK release later this week, can you mark these methods as `@deprecated` instead of deleting them? Otherwise the SDK will lose the capability to query any events since the new `sui_getEvents` endpoint will not be available until the release next week

File: sdk/typescript/src/types/transactions.ts
Comment (666lcz) [inline sdk/typescript/src/types/transactions.ts:183]: This has been defined in https://github.com/MystenLabs/sui/blob/d8633441e79f3d30aba4b001de94808c294b4317/sdk/typescript/src/types/common.ts#L10 ? But there seems to be an outdated line https://github.com/MystenLabs/sui/blob/d8633441e79f3d30aba4b001de94808c294b4317/sdk/typescript/src/types/common.ts#L13 could you help remove it?

File: sdk/typescript/src/providers/provider.ts
Comment (patrickkuo) [reply sdk/typescript/src/providers/provider.ts:200]: Will it be easier if we commit after the next cut?

File: crates/sui-storage/src/event_store/sql.rs
Comment (velvia) [inline crates/sui-storage/src/event_store/sql.rs:331]: Hmm, I wonder if there is a way to avoid all this repetition of the query except for the ASC/DESC part.  Like, we define a single query, but the ASC/DESC can be replaced by something like SEQ_ORDER.  Then a function can replace SEQ_ORDER with ASC/DESC.  This would help prevent accidental mistakes - right now we have no way of ensuring the two queries are the same.

File: crates/sui-storage/src/event_store/mod.rs
Comment (velvia) [inline crates/sui-storage/src/event_store/mod.rs:380]: Can you please update all of the RustDoc's since most of them assume we query by time range, but now we are querying by this cursor.  All the RustDocs should have a section roughly explaining what this cursor is, as it won't be obvious.

File: crates/sui-storage/src/event_store/mod.rs
Comment (velvia) [inline crates/sui-storage/src/event_store/mod.rs:386]: Can we name this parameter `ascending` instead of `reverse`?  I feel this is more obvious.  ascending true means return items in ascending sequence order, but `reverse` it is hard to tell what that means (ie what is normal?)

Comment (velvia) [reply crates/sui-storage/src/event_store/mod.rs:386]: Or rather maybe `descending`.....

File: crates/sui-storage/src/event_store/sql.rs
Comment (velvia) [inline crates/sui-storage/src/event_store/sql.rs:386]: So this is supposed to prevent the same events from being ingested, but it would be better to use an externally assigned sequence number for each event here, instead of tx_seq_num (which repeats for the same events in the same TX).  I know you didn't write this code originally but just saying we should improve this code if possible.

File: crates/sui-storage/src/event_store/sql.rs
Comment (velvia) [inline crates/sui-storage/src/event_store/sql.rs:416]: Ok, so modifying the input events is a really bad idea.  It makes the API non idempotent and its side effects could introduce lots of bugs.  

A second problem is that getting the last inserted rowID for every single row might be a problem when we move to bulk inserts, which I'm likely to do soon for performance reasons.

Instead, can we simply return the last row ID that was generated from the database from the method?    
Also, I'm not sure if we should be feeding back the sequence number from the event store into the other events.  What is the point of that?  The sequence number is kind of an internal thing to implement paging, I'm questioning if it should be an external API.

**Review by velvia — CHANGES_REQUESTED** [review]
Hey @patrickkuo great work.  I only reviewed the event store changes, event handler, event definitions etc.

The biggest issue I have is that in this design, the event sequence number is generated by the event store insert API, then it modifies the events.  Is the intention to have the sequence number broadcasted out for streaming as well?

I think the event sequence number should be generated in event_handler instead of inside a specific implementation, for many reasons.
1. Today there is a hidden dependency this code generates - the event store add() code must be run first, and it must generate the sequence number so that subsequent code can send stuff out with the sequence number in it.  This hidden dependency is hard to figure out and makes the code brittle.
2. As a consequence of 1, for example, I cannot reorder the event ingestion to be in bulk and after the streaming, which I was going to do next... and worse yet this dependency would not be obvious
3. It's a bad idea in APIs to make an input `&mut`.  It allows for arbitrary changes to the input.  We really really don't want to mutate input when possible
4. Generating the sequence number externally makes it easier to use the sequence number to prevent re-ingestion of the same events

Otherwise the changes look great.

File: crates/sui-storage/src/event_store/sql.rs
Comment (patrickkuo) [reply crates/sui-storage/src/event_store/sql.rs:331]: There are no easy way using the SQL's api, I will write a query builder for this to prevent accidental mistakes

File: crates/sui-storage/src/event_store/sql.rs
Comment (patrickkuo) [reply crates/sui-storage/src/event_store/sql.rs:416]: we really only need a unique ID here, the unique id is needed for the pagination cursor so will need to exposes it to the end user, maybe we can use digest or a random hash instead of an incremental index number, WDYT?

Comment (patrickkuo) [reply crates/sui-storage/src/event_store/sql.rs]: I have removed the ID from EventEnvelop, it's now only used for pagination cursor

**Comment by patrickkuo** [comment]
this PR is ready for more reviews, thanks!!

File: crates/sui-gateway/src/unit_tests/rpc_server_tests.rs
Comment (stella3d) [inline crates/sui-gateway/src/unit_tests/rpc_server_tests.rs:391]: Like other large tests, I think this one would be easier to read & debug if we could split it into a test fixture with multiple cases: setup data before all tests, then do each assert as a separate test case in the fixture.

File: crates/sui-storage/src/event_store/sql.rs
Comment (stella3d) [inline crates/sui-storage/src/event_store/sql.rs:441]: putting `cursor` before the primary parameter of the call (`digest` here) instead of with the rest of the queries-organization parameters (for lack of a better term) like limit & order feels off.

I would expect the signature for all these to put `cursor` directly before `limit`.

```db.events_by_transaction(txId, 0, 50, true)```

puts the most relevant data first, compared to
```db.events_by_transaction(0, txId, 50, true)```

**Review by stella3d — CHANGES_REQUESTED** [review]
Only the `Then` -> `Than` change i mentioned that should block merge of this - the rest are mostly about improving readability of tests or method signatures.

File: sdk/typescript/test/e2e/read-events.test.ts
Comment (patrickkuo) [reply sdk/typescript/test/e2e/read-events.test.ts:15]: how about boolean?

File: sdk/typescript/src/types/events.ts
Comment (patrickkuo) [reply sdk/typescript/src/types/events.ts]: after Mark's changes, we are now using "tx_seq":"event_seq" as ID, it is not string instead of number

File: crates/sui-storage/src/event_store/sql.rs
Comment (patrickkuo) [reply crates/sui-storage/src/event_store/sql.rs:1398]: will fix in a follow up PR, this PR is quite big already

File: sdk/typescript/test/e2e/read-events.test.ts
Comment (stella3d) [reply sdk/typescript/test/e2e/read-events.test.ts:15]: @patrickkuo that works great if we can only ever have ascending or descending orders!

File: crates/sui-storage/src/event_store/mod.rs
Comment (patrickkuo) [reply crates/sui-storage/src/event_store/mod.rs:345]: the key of the event entry is now a compound key in the database tx_seq_num + event_seq_num, this 2 number uniquely identify an event entry.

Comment (stella3d) [reply crates/sui-storage/src/event_store/mod.rs:345]: do you think we should add a type alias for this tuple ? might be handy to have that semantic meaning preserved for each element

Comment (patrickkuo) [reply crates/sui-storage/src/event_store/mod.rs]: I have make EventId a struct instead, this should make it much cleaner compare to previous implementation

File: sdk/typescript/test/e2e/read-events.test.ts
Comment (patrickkuo) [reply sdk/typescript/test/e2e/read-events.test.ts]: done!

**Comment by patrickkuo** [comment]
Friendly nudge :) this PR needs an approval.

**Review by velvia — APPROVED** [review]
@patrickkuo great work and thanks for being patient.

I only looked at the event store, event handler and authority RPC methods, and this looks great.

File: crates/sui-json-rpc/src/event_api.rs
Comment (patrickkuo) [reply crates/sui-json-rpc/src/event_api.rs:100]: For this PR yes.
I will be working on complex query next, my goal is to allow user to combine multiple EventQuery objects using And/Or/Not operators to form a more complex query. (same thing will happen to TransactionQuery as well)

the `get_event_query` will become event more busy :| potentially forming a big query string with lot of and/or clauses, will need to make sure this won't cause any memory issues.

**Comment by patrickkuo** [comment]
@stella3d can I get a thumbs up from you?

**Comment by patrickkuo** [comment]
@666lcz do we still have problem lining up TS SDK and Sui release? Do I need to put back the old methods?

File: sdk/typescript/test/e2e/read-transactions.test.ts
Comment (666lcz) [inline sdk/typescript/test/e2e/read-transactions.test.ts:50]: I think using union types `ascending` here is much more readable than a boolean - can we change it back?

**Review by 666lcz — APPROVED** [review]
Great work! Can you address the inline comment and run `pnpm changeset` at the root directory before merging this?

File: sdk/typescript/src/providers/json-rpc-provider.ts
Comment (666lcz) [inline sdk/typescript/src/providers/json-rpc-provider.ts:584]:
```suggestion
      cursor: EventId | null,
```
nit: spacing

File: sdk/typescript/src/providers/json-rpc-provider.ts
Comment (666lcz) [inline sdk/typescript/src/providers/json-rpc-provider.ts:591]:
```suggestion
          [query, cursor, limit, descendingOrder],
```

Sorry just saw @stella3d's previous comment. If we are concerned about byte size for `descendingOrder`, we can replace `descendingOrder` with `order === 'descending'`  here.

**Review by stella3d — APPROVED** [review]

File: sdk/typescript/src/providers/json-rpc-provider.ts
Comment (tam6868) [reply sdk/typescript/src/providers/json-rpc-provider.ts]: Ues

Comment (tam6868) [reply sdk/typescript/src/providers/json-rpc-provider.ts]: Ue

**Pull request merged**
